### ISSUE TITLE
[TIARA] Schema Retriever : Not able to reproduce results.

### ISSUE THREAD

**Issue by PrayushiFaldu** [microsoft/KC#26]
Thank you for the code.

I'm trying to retrain the schema retriever according to README's instructions.
1. For class: I was able to reproduce the results (by changing batch size to 128).
2. For relations: While training the eval loss is decreasing and eval accuracy also reaches around 95%, but prediction on the dev set results in **0** hits@k for all k. (I am not changing any hyperparameters in the code).

Am I missing something? Please help with the issue.

**Comment by yhshu** [comment]
Hi there, could you let me know which dataset you are mentioning? 
For GrailQA, relations work similarly to classes. You may check the relation checkpoint has been trained/loaded, since the class and relation retrieval use different models.
For WebQSP, there's no class in S-expression, and the candidate relations are obtained from the entity neighborhood. So we use a cache here to store candidate relations, and it should be loaded before training/evaluation.

**Comment by PrayushiFaldu** [comment]
I did for GrailQA.
Since I am re-training so I first deleted pre-trained models from the directory and then run the following commands:
python retriever/dense_retriever/grailqa_schema_dense_retriever.py --schema_type class --split dev
python retriever/dense_retriever/grailqa_schema_dense_retriever.py --schema_type relation --split dev
From eval loss and accuracy I can see the model trains correctly (for relation) but is not able to predict.

**Comment by yhshu** [comment]
I just re-run the evaluation with the checkpoint, and it works.
So I think the training should be the problem. Eval loss and accuracy cannot accurately tell if the model has been trained correctly. Would you please check if the `encode` function has saved reasonable training data?

**Comment by PrayushiFaldu** [comment]
How do we define reasonable data?
I checked the file size and sample data. They seem fine. And I am using the same script for class and relation, if the issue is with the saving of training data then it should happen with both (relation and class) right?
Can you please try re-training at your end once?

**Comment by yhshu** [comment]
I'm talking about whether the value of three lists `text_a` `text_b` and `labels` in that function are matched in your program.
I will follow up to check.

**Comment by PrayushiFaldu** [comment]
Yes, the values of the variables text_a, text_b, and labels in the code match the original training data.

**Comment by yhshu** [comment]
Re-training works fine for me. In this case, I suggest that there are other issues to check:
- evaluate the model on the dev set rather than the test set
- make sure to use the relation retrieval model instead of the class retrieval model

**Comment by PrayushiFaldu** [comment]
Ok, I will check. Thanks.
